mostlyai/sdk/domain.py

```diff
@@ -3408,6 +3443,22 @@ def add_model_configuration(self):
             self.language_model_configuration.max_sequence_window = None
         return self
 
+    @model_validator(mode="after")
+    def extract_columns_from_data_if_missing(self):
+        """extract column names from base64 data if columns are not provided."""
+        if self.columns is None and self.data is not None and len(self.data) >= 512:
+            # assume base64-encoded parquet if data is long enough
+            try:
+                df = convert_to_df(self.data, format="parquet")
+                self.columns = [
+                    SourceColumnConfig(name=col_name, model_encoding_type=ModelEncodingType.auto)
+                    for col_name in df.columns
+                ]
+            except Exception:
+                # if conversion fails, leave columns as None
+                pass
+        return self
```

Comment:
Couldn't follow, why we need this validator?
My guess its for `validate_constraints`. But if `validate_constraints` requires columns, we can't guarantee them here, as data might not be passed explicitly, but by connector.